File: packages/android-im-ui/src/test/kotlin/com/flare/im/ui/ConversationHeaderGroupingTest.kt
package com.flare.im.ui

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertTrue

class ConversationHeaderGroupingTest {
    private val group = ConversationIdentity("g1", "Product room", ConversationHeaderKind.Group)

    @Test
    fun headerActionsResolveDefaultsCapabilitiesAndHostConfiguration() {
        val resolved = resolveConversationHeaderActions(
            identity = group,
            capabilities = ConversationHeaderCapabilities(setOf("search", "addMember", "share", "task", "pin")),
            configuration = ConversationHeaderConfiguration(
                removeActionIds = setOf("share"),
                actionOverrides = listOf(
                    ConversationHeaderAction(
                        id = "addMember",
                        label = "Invite people",
                        icon = "person-add",
                        placement = ConversationHeaderActionPlacement.Add,
                        order = 30,
                        enabled = false,
                        disabledReason = "Read only",
                    ),
                ),
            ),
            actions = listOf(
                ConversationHeaderAction("pin", "Pin", placement = ConversationHeaderActionPlacement.Primary, order = 5),
                ConversationHeaderAction("task", "Create task", placement = ConversationHeaderActionPlacement.Add, order = 40),
            ),
        )
        assertEquals(listOf("pin", "search", "addMember", "task"), resolved.map { it.id })
        assertFalse(resolved[2].enabled)
        assertEquals("Invite people", resolved[2].label)
    }

    @Test
    fun groupingCoversSenderTimeAndSystemBoundaries() {
        val messages = listOf(
            message("1", "ivy", 1_000), message("2", "ivy", 2_000),
            message("3", "ivy", 400_000), message("4", "system", 401_000, system = true),
            message("5", "ivy", 402_000), message("6", "me", 403_000),
        )
        assertEquals(MessageGroupPosition.First, messageGroupPosition(messages, 0))
        assertEquals(MessageGroupPosition.Last, messageGroupPosition(messages, 1))
        assertEquals(MessageGroupPosition.Single, messageGroupPosition(messages, 2))
        assertEquals(MessageGroupPosition.Single, messageGroupPosition(messages, 3))
        assertEquals(MessageGroupPosition.Single, messageGroupPosition(messages, 4))

        // A recalled message is a notice: it stands alone and ends the sender's run.
        val recalledRun = listOf(
            message("r1", "ivy", 1_000), message("r2", "ivy", 2_000),
            message("r3", "ivy", 3_000).copy(lifecycle = FlareMessageLifecycle(mutation = FlareMessageMutationState.Recalled)),
            message("r4", "ivy", 4_000),
        )
        assertEquals(
            listOf(MessageGroupPosition.First, MessageGroupPosition.Last, MessageGroupPosition.Single, MessageGroupPosition.Single),
            recalledRun.indices.map { messageGroupPosition(recalledRun, it) },
        )

        val incoming = messageRowPresentation(messages.first(), MessageGroupPosition.First, "me", true)
        val outgoing = messageRowPresentation(messages.last(), MessageGroupPosition.Single, "me", true)
        assertTrue(incoming.showAvatar)
        assertTrue(incoming.showSenderName)
        assertTrue(incoming.reserveAvatarSpace)
        assertFalse(outgoing.showAvatar)
        assertFalse(outgoing.showSenderName)
    }

    private fun message(id: String, sender: String, time: Long, system: Boolean = false) = FlareMessageData(
        id = id,
        senderId = sender,
        senderName = sender,
        sentAtMs = time,
        content = if (system) FlareNotificationContent("boundary") else FlareTextContent("hello"),
    )
}

/** The More button performs the only overflow action itself; two or more keep the menu. */
class ConversationHeaderOverflowTest {
    private val direct = ConversationIdentity("u1", "陈默")

    @Test fun oneOverflowActionTakesTheMoreButton() {
        // Compact phones keep one primary button: search stays, and details alone overflows.
        val resolved = resolveConversationHeaderActions(
            direct,
            configuration = ConversationHeaderConfiguration(removeActionIds = setOf("audioCall", "videoCall", "share")),
        )
        val layout = conversationHeaderLayout(resolved, maxPrimary = 1)
        assertEquals(listOf("search"), layout.primary.map { it.id })
        assertEquals("details", layout.soleOverflow?.id)
    }

    @Test fun twoOrMoreOverflowActionsKeepTheMenu() {
        val resolved = resolveConversationHeaderActions(direct, configuration = ConversationHeaderConfiguration(removeActionIds = setOf("share")))
        val compact = conversationHeaderLayout(resolved, maxPrimary = 1)
        assertEquals(listOf("audioCall", "videoCall", "details"), compact.overflow.map { it.id })
        assertEquals(null, compact.soleOverflow)
        // With room for every primary action only details overflows, and it takes the More button again.
        assertEquals("details", conversationHeaderLayout(resolved, maxPrimary = 3).soleOverflow?.id)
        // Nothing overflows: there is no More button at all.
        val withoutDetails = resolveConversationHeaderActions(direct, configuration = ConversationHeaderConfiguration(removeActionIds = setOf("share", "details")))
        assertEquals(emptyList(), conversationHeaderLayout(withoutDetails, maxPrimary = 3).overflow)
    }
}

File: packages/flutter-im-ui/test/conversation_header_grouping_test.dart
import 'package:flare_im_ui/flare_im_ui.dart';
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

FlareMessageData message(
  String id,
  String sender,
  int time, {
  bool system = false,
}) => FlareMessageData(
  id: id,
  senderId: sender,
  senderName: sender,
  sentAtMs: time,
  content: system
      ? const FlareNotificationContent('boundary')
      : const FlareTextContent('hello'),
);

void main() {
  const group = FlareConversationIdentity(
    id: 'g1',
    title: 'Product room',
    kind: FlareConversationHeaderKind.group,
  );

  test(
    'header actions resolve defaults, capabilities and host configuration',
    () {
      final resolved = resolveConversationHeaderActions(
        identity: group,
        capabilities: const FlareConversationHeaderCapabilities(
          availableActionIds: {'search', 'addMember', 'share', 'task', 'pin'},
        ),
        configuration: const FlareConversationHeaderConfiguration(
          removeActionIds: {'share'},
          actionOverrides: [
            FlareConversationHeaderAction(
              id: 'addMember',
              label: 'Invite people',
              icon: 'person-add',
              placement: FlareConversationHeaderActionPlacement.add,
              order: 30,
              enabled: false,
              disabledReason: 'Read only',
            ),
          ],
        ),
        actions: const [
          FlareConversationHeaderAction(
            id: 'pin',
            label: 'Pin',
            placement: FlareConversationHeaderActionPlacement.primary,
            order: 5,
          ),
          FlareConversationHeaderAction(
            id: 'task',
            label: 'Create task',
            placement: FlareConversationHeaderActionPlacement.add,
            order: 40,
          ),
        ],
      );

      expect(resolved.map((action) => action.id), [
        'pin',
        'search',
        'addMember',
        'task',
      ]);
      expect(resolved[2].enabled, isFalse);
      expect(resolved[2].label, 'Invite people');
    },
  );

  testWidgets('compact header moves excess primary actions into More', (
    tester,
  ) async {
    await tester.pumpWidget(
      const MaterialApp(
        home: Scaffold(
          body: SizedBox(
            width: 360,
            child: FlareConversationHeader(identity: group, onAction: _ignore),
          ),
        ),
      ),
    );
    expect(find.text('Product room'), findsOneWidget);
    expect(find.byIcon(Icons.add_rounded), findsOneWidget);
    expect(find.byIcon(Icons.more_horiz_rounded), findsOneWidget);
  });

  test('message grouping covers sender, time and system boundaries', () {
    final messages = [
      message('1', 'ivy', 1000),
      message('2', 'ivy', 2000),
      message('3', 'ivy', 400000),
      message('4', 'system', 401000, system: true),
      message('5', 'ivy', 402000),
      message('6', 'me', 403000),
    ];
    expect(
      flareMessageGroupPosition(messages, 0),
      FlareMessageGroupPosition.first,
    );
    expect(
      flareMessageGroupPosition(messages, 1),
      FlareMessageGroupPosition.last,
    );
    expect(
      flareMessageGroupPosition(messages, 2),
      FlareMessageGroupPosition.single,
    );
    expect(
      flareMessageGroupPosition(messages, 3),
      FlareMessageGroupPosition.single,
    );
    expect(
      flareMessageGroupPosition(messages, 4),
      FlareMessageGroupPosition.single,
    );

    // A recalled message is a notice: it stands alone and ends the sender's run.
    final recalledRun = [
      message('r1', 'ivy', 1000),
      message('r2', 'ivy', 2000),
      FlareMessageData(
        id: 'r3',
        senderId: 'ivy',
        senderName: 'ivy',
        sentAtMs: 3000,
        content: const FlareTextContent('gone'),
        lifecycle: const FlareMessageLifecycle(
          mutation: FlareMessageMutationState.recalled,
        ),
      ),
      message('r4', 'ivy', 4000),
    ];
    expect(
      [
        for (var i = 0; i < recalledRun.length; i++)
          flareMessageGroupPosition(recalledRun, i),
      ],
      [
        FlareMessageGroupPosition.first,
        FlareMessageGroupPosition.last,
        FlareMessageGroupPosition.single,
        FlareMessageGroupPosition.single,
      ],
    );

    final incoming = flareMessageRowPresentation(
      message: messages.first,
      position: FlareMessageGroupPosition.first,
      currentUserId: 'me',
      groupConversation: true,
    );
    final outgoing = flareMessageRowPresentation(
      message: messages.last,
      position: FlareMessageGroupPosition.single,
      currentUserId: 'me',
      groupConversation: true,
    );
    expect(incoming.showAvatar, isTrue);
    expect(incoming.showSenderName, isTrue);
    expect(incoming.reserveAvatarSpace, isTrue);
    expect(outgoing.showAvatar, isFalse);
    expect(outgoing.showSenderName, isFalse);
  });
}

void _ignore(FlareConversationHeaderAction action) {}
